### PR TITLE
Fix and test edge cases in handling of AWS4 headers

### PR DESCRIPTION
**Pull request by ararslan**
In particular, we were not correctly handling duplicate and multiline headers.

### CODE REVIEW

File: src/AWS4AuthRequest.jl [hunk 1]
```diff
@@ -34,6 +34,17 @@ function request(::Type{AWS4AuthLayer{Next}},
     return request(Next, url, req, body; kw...)
 end
 
+# Normalize whitespace to the form required in the canonical headers.
+# Note that the expected format for multiline headers seems not to be explicitly
+# documented, but Amazon provides a test case for it, so we'll match that behavior.
+function _normalize_ws(s::AbstractString)
```
Comment (oxinabox) [inline src/AWS4AuthRequest.jl]: Can we just fully spell that out:
```
_normalize_whitespace
```

Comment (ararslan) [reply src/AWS4AuthRequest.jl]: It's an internal function used in one place and I have explanatory comments at the definition and at the call site, so that seems unnecessary to me.

File: src/AWS4AuthRequest.jl [hunk 2]
```diff
@@ -90,10 +101,14 @@ function sign_aws4!(method::String,
     end
 
     # Sort and lowercase() Headers to produce canonical form...
-    canonical_headers = sort!(map(headers) do (k, v)
-        string(lowercase(k), ':', replace(strip(v), r"\s+" => " "))
-    end)
-    signed_headers = join(sort!(map(lowercase∘first, headers)), ";")
+    unique_headers = sort!(unique!(map(lowercase∘first, headers)))
```
Comment (oxinabox) [inline src/AWS4AuthRequest.jl]: `unique_headers_keys`

File: src/AWS4AuthRequest.jl [hunk 3]
```diff
@@ -90,10 +101,14 @@ function sign_aws4!(method::String,
...
+    canonical_headers = map(unique_headers) do k
+        # All values present for this key (case-insensitive) with whitespace normalized
+        vs = map(i->_normalize_ws(last(headers[i])),
+                 findall(x->lowercase(first(x)) == k, headers))
+        string(lowercase(k), ':', join(vs, ','))
```
Comment (oxinabox) [inline src/AWS4AuthRequest.jl]:
```suggestion
        string(k, ':', join(vs, ','))
```

`k` is alrady lowercase

File: src/AWS4AuthRequest.jl [hunk 4]
```diff
@@ -90,10 +101,14 @@ function sign_aws4!(method::String,
...
     # Sort and lowercase() Headers to produce canonical form...
-    canonical_headers = sort!(map(headers) do (k, v)
-        string(lowercase(k), ':', replace(strip(v), r"\s+" => " "))
-    end)
-    signed_headers = join(sort!(map(lowercase∘first, headers)), ";")
+    unique_headers = sort!(unique!(map(lowercase∘first, headers)))
+    canonical_headers = map(unique_headers) do k
+        # All values present for this key (case-insensitive) with whitespace normalized
+        vs = map(i->_normalize_ws(last(headers[i])),
+                 findall(x->lowercase(first(x)) == k, headers))
```
Comment (oxinabox) [inline src/AWS4AuthRequest.jl]: This map + findall is kind scary
I think just writing out the for loop is cleaner

Something like: 
```
vs = Vector{Header}() # is this eltype right?
for header in headers
	header_key = lowercase(first(header))
	if header_key = k
		header_value
		push!(vs, _normalize_ws(header_value))
	end
end
```

File: src/AWS4AuthRequest.jl (hunk 1 above)
Comment (oxinabox) [reply src/AWS4AuthRequest.jl]: The comment at the callsite wouldn't be required if it wasn't true that `ws` is a unclear acronym.

File: src/AWS4AuthRequest.jl [hunk 5]
```diff
@@ -34,6 +34,17 @@ function request(::Type{AWS4AuthLayer{Next}},
...
 
+# Normalize whitespace to the form required in the canonical headers.
+# Note that the expected format for multiline headers seems not to be explicitly
+# documented, but Amazon provides a test case for it, so we'll match that behavior.
+function _normalize_ws(s::AbstractString)
+    if any(isequal('\n'), s)
+        join(map(_normalize_ws, split(s, '\n')), ',')
+    else
+        replace(strip(s), r"\s+" => " ")
+    end
```
Comment (omus) [inline src/AWS4AuthRequest.jl]: There seems to be some subtleties that are worth commenting about. So you want to replace each `\n` with a `,` and also remove all whitespace around the newlines. Any remaining contiguous whitespace is replaced with a single space.

File: src/AWS4AuthRequest.jl [hunk 6]
```diff
@@ -90,10 +101,14 @@ function sign_aws4!(method::String,
...
-        string(lowercase(k), ':', replace(strip(v), r"\s+" => " "))
-    end)
-    signed_headers = join(sort!(map(lowercase∘first, headers)), ";")
+    unique_header_keys = sort!(unique!(map(lowercase∘first, headers)))
+    canonical_headers = map(unique_header_keys) do k
+        # All values present for this key (case-insensitive) with whitespace normalized
+        vs = map(i->_normalize_ws(last(headers[i])),
+                 findall(x->lowercase(first(x)) == k, headers))
+        string(k, ':', join(vs, ','))
+    end
```
Comment (omus) [inline src/AWS4AuthRequest.jl]: Alternative:
```julia
using DataStructures: DefaultOrderedDict
normalized_headers = DefaultOrderedDict{String,Vector{String}}(() -> String[])
for k in sort!(collect(keys(headers)))
    ck = lowercase(k)
    push!(canonical_headers[ck], _normalize_ws(header[k]))
end
canonical_headers = [string(k, ':', join(vs, ',')) for (k, vs) in normalized_headers]
unique_header_keys = collect(keys(normalized_headers))
```
or for dependency haters:
```julia
unique_header_keys = Vector{String}()
normalized_headers = Dict{String,Vector{String}}()
for k in sort!(collect(keys(headers)))
    ck = lowercase(k)

    if !haskey(canonical_headers, ck)
        canonical_headers[ck] = Vector{String}()
    end

    push!(canonical_headers[ck], _normalize_ws(header[k]))
    push!(normalized_headers, ck)
end
canonical_headers = [string(k, ':', join(vs, ',')) for (k, vs) in normalized_headers]
```
The main advantage to this approach is you avoid some iterations

File: test/aws4.jl [hunk 7]
```diff
@@ -6,38 +6,84 @@ using HTTP.AWS4AuthRequest: sign_aws4!
...
+               include_sha256=false,
+               opts...)
+    headers
+end
+
+test_auth_string(headers, sig) =
+    string("AWS4-HMAC-SHA256 ",
+           "Credential=AKIDEXAMPLE/20150830/us-east-1/service/aws4_request, ",
+           "SignedHeaders=", headers, ", ",
+           "Signature=", sig)
```
Comment (omus) [inline test/aws4.jl]: Hard to read:
```julia
function test_auth_string(headers, sig) 
    d = [
        "AWS4-HMAC-SHA256 Credential" => "AKIDEXAMPLE/20150830/us-east-1/service/aws4_request",
        "SignedHeaders" => headers,
        "Signature" => sig,
    ]
    join(("$k=$v" for (k, v) in d), ", ")
end
```

File: test/aws4.jl [hunk 8]
```diff
@@ -6,38 +6,84 @@ using HTTP.AWS4AuthRequest: sign_aws4!
...
-                              "Signature=", sig)
-                @test d["Authorization"] == auth
-            end
+    noheaders = [
+        ("get-vanilla", "", "5fa00fa31553b73ebf1942676e86291e8372ff2a2260956d9b8aae1d763fbf31"),
+        ("get-vanilla-empty-query-key", "?Param1=value1", "a67d582fa61cc504c4bae71f336f98b97f1ea3c7a6bfe1b6e45aec72011b9aeb"),
+        ("get-utf8", "ሴ", "8318018e0b0f223aa2bbf98705b62bb787dc9c0e678f255a891fd03141be5d85"),
+    ]
+    for (name, p, sig) in noheaders
+        @testset "$name" begin
```
Comment (omus) [inline test/aws4.jl]: Note that: `@testset "$name" for (name, p, sig) in noheaders` is a thing

File: src/AWS4AuthRequest.jl (hunk 6 above)
Comment (ararslan) [reply src/AWS4AuthRequest.jl]: I had to revert this change, as it did not produce the correct result.

Comment (omus) [reply src/AWS4AuthRequest.jl]: All is well now

**Comment by ararslan** [comment]
The hashes produced on 32-bit Windows seem to be different. I am _very_ thoroughly annoyed by this,

**Comment by ararslan** [comment]
I can reproduce in a 32-bit Linux VM. Time to investigate...

**Comment by ararslan** [comment]
64-bit Windows failure with Julia nightly seems to be a network issue.
